src/array/struct_/mutable.rs

```diff
@@ -38,44 +39,77 @@ impl MutableStructArray {
         Self::from_data(data_type, values, None)
     }
 
-    /// Create a [`MutableStructArray`] out of low-end APIs.
-    /// # Panics
-    /// This function panics iff:
+    /// Fallibly create a [`MutableStructArray`] out of low-level APIs.
+    /// # Errors
+    /// This function returns an error if:
     /// * `data_type` is not [`DataType::Struct`]
     /// * The inner types of `data_type` are not equal to those of `values`
     /// * `validity` is not `None` and its length is different from the `values`'s length
-    pub fn from_data(
+    pub fn try_from_data(
```

Comment:
could we rename it to `try_new` instead? It is more aligned with the rest of the containers